### DISCUSSION TITLE
Hig Precision and IMU dead reckoning

### DISCUSSION THREAD

**Discussion by HackDiverseRam** [semuconsulting/pygnssutils#95] [Q&A]
Hey everyone,

I am currently using a TW5790 GPS that has a built in IMU. I can obtain base NMEA messages from this GPS but I want to build a python script that can initialize the NMEA High precision feature. If possible I would love to know more about how to set up the IMU dead reckoning feature as well. I'm still new to utilizing GPS in such a way and would appreciate any help on the matter.

**Comment by semuadmin** [comment]
Hi @HackDiverseRam,

This is probably more a question for the [u-blox Support Community](https://portal.u-blox.com/s/), but I'll endeavour to provide some basic pointers.

If I understand correctly, the TW5790 is based on the u-blox ZED-F9R receiver, so your first port of call should be the [Interface](https://www.u-blox.com/sites/default/files/documents/u-blox-F9-HPS-1.30_InterfaceDescription_UBX-22010984.pdf) and [Integration](https://www.u-blox.com/sites/default/files/ZED-F9R_Integrationmanual_UBX-20039643.pdf) Manuals for that particular receiver - specifically the sections in the Integration Manual (section 3.2.4 onwards) that refer to IMU configuration and calibration. I should warn you - it is NOT a quick or trivial process! pygnssutils and its underlying pyubx2 library support all the necessary UBX configuration commands, but they do not provide any kind of 'quick fix' - you'll need to do the research for your particular use case.

Enabling NMEA high precision mode is more straightforward - you'll need to create a UBX CFG-VALSET message with the configuration database key CFG_NMEA_HIGHPREC (0x10930006) set to 1, and send this to the receiver. Refer to the Interface Manual (section 6) and the [pyubx2 README](https://github.com/semuconsulting/pyubx2?tab=readme-ov-file#configinterface) for details on how to do this - you'll need something like this (*substitute the serial port as required*):

```python
from serial import Serial
from pyubx2 import UBXMessage, SET_LAYER_RAM, TXN_NONE

with Serial("/dev/ttyACM1", 38400, timeout=3) as stream:
    msg = UBXMessage.config_set(SET_LAYER_RAM, TXN_NONE, [("CFG_NMEA_HIGHPREC", 1)])
    stream.write(msg.serialize())
    print("config command written to receiver")

```

Alternatively you could switch to the UBX protocol (rather than using NMEA) and enable the UBX NAV-HPPOSLLH (high precision LLH position) or UBX NAV-HPPOSECEF (high precision ECEF position) message types.
